Source/WebCore/platform/audio/SharedAudioDestination.cpp

```diff
@@ -0,0 +1,315 @@
+    return map;
+}
+
+Ref<SharedAudioDestinationAdapter> SharedAudioDestinationAdapter::ensureAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction)
+{
+    auto key = std::make_tuple(numberOfOutputChannels, sampleRate);
+    auto results = sharedMap().find(key);
+    if (results != sharedMap().end()) {
+        if (auto existingAdapter = results->value.get())
+            return existingAdapter.releaseNonNull();
+    }
+
+    Ref<SharedAudioDestinationAdapter> newAdapter = adoptRef(*new SharedAudioDestinationAdapter(numberOfOutputChannels, sampleRate, WTFMove(ensureFunction)));
+    auto weakAdapter = ThreadSafeWeakPtr<SharedAudioDestinationAdapter> { newAdapter.get() };
+    sharedMap().set(key, WTFMove(weakAdapter));
+    return newAdapter;
+}
+
+SharedAudioDestinationAdapter::SharedAudioDestinationAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction)
+    : m_numberOfOutputChannels { numberOfOutputChannels }
+    , m_sampleRate { sampleRate }
+    , m_destination { ensureFunction(*this) }
+    , m_workBus { AudioBus::create(numberOfOutputChannels, AudioUtilities::renderQuantumSize).releaseNonNull() }
+    , m_configurationQueue { WorkQueue::create("SharedAudioDestinationAdapter configuration queue") }
+    , m_configurationSemaphore(0)
+{
+}
+
+SharedAudioDestinationAdapter::~SharedAudioDestinationAdapter()
+{
+    auto key = std::make_tuple(m_numberOfOutputChannels, m_sampleRate);
+    sharedMap().remove(key);
+    m_configurationQueue->dispatchSync([] { });
+    m_destination->clearCallback();
+}
+
+void SharedAudioDestinationAdapter::addRenderer(SharedAudioDestination& renderer, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_configurationQueue->dispatch([this, weakThis = ThreadSafeWeakPtr { *this }, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
+        auto protectedThis = weakThis.get();
+        if (!protectedThis)
+            return;
+
+        if (!m_renderers.contains(renderer))
+            m_renderers.append(WTFMove(renderer));
+        configureRenderThread(WTFMove(completionHandler));
+    });
+}
+
+void SharedAudioDestinationAdapter::removeRenderer(SharedAudioDestination& renderer, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_configurationQueue->dispatch([this, weakThis = ThreadSafeWeakPtr { *this }, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
+        auto protectedThis = weakThis.get();
+        if (!protectedThis)
+            return;
+
+        m_renderers.removeFirst(renderer);
+        configureRenderThread(WTFMove(completionHandler));
+    });
+}
+
+void SharedAudioDestinationAdapter::configureRenderThread(CompletionHandler<void(bool)>&& completionHandler)
+{
+    ASSERT(m_configurationQueue->isCurrent());
+    {
+        LockHolder locker { m_renderLock };
```

Comment:
`LockHolder` is useless in new code, now that we support C++17. We should likely get rid of it. You can just use `Locker` here, which is shorter and equivalent (since `LockHolder` is an alias to `Locker<Lock>` and the template parameter gets deduced with C++17).